keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,787 @@
+
+See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.
+
+### Notes/Constraints/Caveats (Optional)
+
+A focus of this KEP is for accounting for deletion via disruptions (ie kubelet eviction, preemption, etc) so we want to mention some other cases where the job controller can start deletion.
+
+1) A job is over the `activeDeadlineSeconds` so the child pods are deleted.  
+2) With `PodFailurePolicy` active and `FailJob` is set as the action, the pods that are still running would be deleted.
+
+### Risks and Mitigations
+
+One area of contention is how this KEP will work with [3329-retriable-and-non-retriable-failures](https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md).  It is important to mention the subtleties here. 
+
+In 3329, there was a decision to make kubelet transition pods to failed before deleting them.  This is feature toggled guarded by `PodDisruptionCondition`.  
+This means that when this feature is turned on, the job controller can count pods as failed once they are fully terminated.  This means that the pod is still considered running, as opposed to failed.  So this pod would be considered active until it is fully terminated.  
+If `PodDisruptionCondition` is turned off then the job controller considers the pod as failed as soon as it is terminating (has a deletion timestamp), because there is no guarantee that the pod will transition to phase=Failed.
+
+Another issue is described [here](https://github.com/kubernetes/enhancements/pull/3940#discussion_r1180777509).
+If PodDisruptionConditions is disabled, a pod bound to a no-longer-existing node may be stuck in the Running phase. As a consequence, it will never be replaced, so the whole job will be stuck from making progress.  Due to the above issue, we can set phase=Failed when PodDisruptionConditions is enabled OR JobRecreatePodsWhenFailed is enabled. When JobRecreatePodsWhenFailed enabled, but PodDisruptionConditions disabled we would just set the phase, but without adding the condition.  We will need to modify the PodGC (gc_controller.go) in the case of `JobRecreatePodsFailed` being enabled while `PodDisruptionConditions` is disabled.  
+
+## Design Details
+
+### Job API Definition
+
+At the JobSpec level, we are adding a new enum field:
+
+```golang
+// This field controls when we recreate pods
+// Default will be TerminatingOrFailed ie recreate pods when they are failed
+// +enum 
+type RecreatePodsWhen string
+const (
+ // This is a field that recreates pods when they are marked as terminating or failed
+ // ie this means that as soon as pods get marked for deletion they will be recreated
+ TerminatingOrFailed RecreatePodsWhen = "TerminatingOrFailed"
```

Comment:
How does this interact with the `PodFailurePolicy` today?  is it
1. pod gets a deletiontimestamp and then say a grace period of 10 minutes to shutdown.
2. job controller marks as failed and creates new pod
3. PodFailurePolicy somehow rewrites the failure after 10 minutes?

Reply:
When using PodFailurePolicy, we still need to wait for the pod to terminate, because the policy might require an exit code. In other words, the current behavior is equivalent `recreatePodsWhen: Failed` when the user sets a `podFailurePolicy`.

Reply:
> Edit: The behavior is that the job controller immediately creates a replacement Pod, even though the previous pod is not counted as "failed" yet.

I'm still confused then. Does this mean that the `PodFailurePolicy` is ineffective in the default case if the termination grace period is long enough for the controller to take action before the pod has an exit code?

Reply:
To some degree.
The job controller would still wait for the Pod to finish, and take action based on it.
If the action is "FailJob", then the job controller would still fail the job and delete the replacement pod that it created earlier.

Reply:
I create this

```yaml
spec:
  backoffLimit: 0
  recreatePodsWhen: nil
  podFailurePolicy:
    rules:
    - action: Ignore
      onExitCodes: 
        operator: In
        values:
        - 127
# podspec with 10 minute graceful termination
```

With defaulting, I get back
```yaml
spec:
  backoffLimit: 0
  recreatePodsWhen: TerminatingOrFailed
  podFailurePolicy:
    rules:
    - action: Ignore
      onExitCodes: 
        operator: In
        values:
        - 127
# podspec with 10 minute graceful termination
```

If the pod is deleted (deletionTimestamp != nil), then the pod is clearly terminating, but it may still exit with code 127, we don't know yet. If the pod is considered failed by the job, then no replacement pod should be created and the job should be failed.  But if the pod exits with 127, then the failure should be ignored and a new pod created.  If the controller honors the TerminatingOrFailed, then the doc says it will be created, but this would violate the backoffLimit.  If the controller honors the podFailurePolicy and waits for the exit code, then the recreatePodsWhen was not honored.

What is proposed to happen?

Reply:
This is one of those tricky scenarios. Because you are using `podFailurePolicy`, the Pod is not counted towards `active` nor `failed`. Under this perspective, the pod recreation is legit.

If you use the new `recreatePodsWhen: Failed`, then we will:
- respect backoffLimit based on `failed`.
- track Pods that are "active or terminating". We would hold recreation based on this tracker, independently of the value of `active`.

One alternative is to consider this a bug of `podFailurePolicy` (still in Beta) and always track "active or terminating". Then this means that when using `podFailurePolicy` the default is `recreatePodsWhen: Failed` and you cannot use `recreatePodsWhen: TerminatingOrFailed`.